gadget/mountedfilesystem.go

```diff
@@ -186,3 +205,487 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+		return fmt.Errorf("cannot backup directory %q: %v", source, err)
+	}
+
+	fis, err := ioutil.ReadDir(srcName)
+	if err != nil {
+		return fmt.Errorf("cannot list directory entries: %v", err)
+	}
+
+	if !strings.HasSuffix(source, "/") {
+		// /foo -> /bar/ => /bar/foo
+		target = filepath.Join(target, filepath.Base(source))
+	}
+
+	for _, fi := range fis {
+		pSrc := filepath.Join(source, fi.Name())
+		pDst := filepath.Join(target, fi.Name())
+
+		backup := f.backupOrCheckpointFile
+		if fi.IsDir() {
+			if err := f.backupOrCheckpointPrefix(dstRoot, pDst+"/", backupDir); err != nil {
+				return fmt.Errorf("cannot backup prefix directory: %v", err)
+			}
+
+			backup = f.backupOrCheckpointDirectory
+			pSrc += "/"
+		}
+		if err := backup(dstRoot, pSrc, pDst, preserveInDst, backupDir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointPrefix(dstRoot, target string, backupDir string) error {
+
+	// file is different and backed up, check how much of the prefix needs to be created
+	for prefix := filepath.Dir(target); prefix != "." && prefix != "/"; prefix = filepath.Dir(prefix) {
+		_, prefixDst, prefixBackupBase := f.entryNames(dstRoot, "", prefix, backupDir)
+
+		prefixBackupName := prefixBackupBase + ".backup"
+		if osutil.FileExists(prefixBackupName) {
+			continue
+		}
+		if !osutil.IsDirectory(prefixDst) {
+			// does not exist now, will be created on the fly and
+			// removed during rollback
+			continue
+		}
+		if err := os.MkdirAll(filepath.Dir(prefixBackupName), 0755); err != nil {
+			return fmt.Errorf("cannot create backup prefix: %v", err)
+		}
+		if err := makeStamp(prefixBackupName); err != nil {
+			return fmt.Errorf("cannot create a checkpoint for directory: %v", err)
+		}
+	}
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointFile(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+
+	srcName, dstName, backupBase := f.entryNames(dstRoot, source, target, backupDir)
+
+	backupName := backupBase + ".backup"
+	sameStamp := backupBase + ".same"
+	preserveStamp := backupBase + ".preserve"
+
+	if osutil.FileExists(backupName) || osutil.FileExists(sameStamp) {
+		// file already checked, move on
+		return nil
+	}
+
+	if strutil.SortedListContains(preserveInDst, dstName) {
+		if !osutil.FileExists(dstName) {
+			// preserve, but does not exist, will be deployed anyway
+			return nil
+		}
+		if osutil.FileExists(preserveStamp) {
+			// already stamped
+			return nil
+		}
+		// make a stamp
+		if err := makeStamp(preserveStamp); err != nil {
+			return fmt.Errorf("cannot create preserve stamp: %v", err)
+		}
+		return nil
+	}
+
+	orig, err := os.Open(dstName)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot open destination file: %v", err)
+		}
+
+		// destination does not exist, no need for backup
+		return f.backupOrCheckpointPrefix(dstRoot, target, backupDir)
```

Comment:
why do we need this call? shouldn't the prefix be taken care by the caller?  the comment is also strange saying no backup and then calling a function with backup in the name.

I honestly find this code hard to follow.

Reply:
Fair point. I moved back the prefix checkpoint to relevant callers.